### PR TITLE
Add augmentor support

### PR DESCRIPTION
**Pull request by ruimashita**
Add augmentation support.

* Add augmentation selector in `init` CLI.
* Documentation
  * Add augmentation reference page in docs site. preview: https://deploy-preview-14--blueoil.netlify.com/reference/data_augmentor.html
  * Describe how to set augmentation in generated yml. preview: https://deploy-preview-14--blueoil.netlify.com/usage/init.html#data-augmentation
* Remove `is_bounding_box` parameter from `FlipLeftRight` and `FlipTopBottom`, and edit these classes to work well even the task is classification or object detection.  Delete `is_bounding_box` from config.yml under`dlk/examples` dirs also.
* Remove default `size` parameter from Crop, because the `size` is should be decided by user.
* Edit `init` unit test in `blueoil_test.sh` that `enable data augmentation` is selecting Blur case .
* Edit `tests/config/make_yml_config` to create test template.

Take over from #2

### CODE REVIEW

**Comment by ruimashita** [comment]
@lm-jira @nlpng 
Review please 👍

File: dlk/examples/classification/lmnet_quantize_cifar10/config.yaml [hunk 1]
```diff
@@ -24,7 +24,6 @@ DATASET:
       is_resize: false
       width: 32
     - !!python/object:lmnet.data_augmentor.FlipLeftRight
-      is_bounding_box: false
```
Comment (lm-jira) [inline dlk/examples/classification/lmnet_quantize_cifar10/config.yaml]: I just wonder is it okay to remove `is_bounding_box` line without removing `probability` line?
Why not removing both of them?

Comment (ruimashita) [reply dlk/examples/classification/lmnet_quantize_cifar10/config.yaml]: @lm-jira
I removed only `is_bounding_box` parameter, don’t remove `probability` parameter. `probability` still remain.

**Comment by lm-jira** [comment]
@ruimashita 
After trying to train the model, there is an error said that
> `FlipLeftRight(probability=0.5, ),` contains invalid syntax.

When I don't apply that augmentor, there is another error said that 
> `])` contains invalid syntax.
`])` is probably the ending bracket of the `DATA_AUGMENTOR` sequence.

**Comment by ruimashita** [comment]
@lm-jira Thanks, I will try to fix it.

**Comment by ruimashita** [comment]
@lm-jira 
When I use bellow config, the error didn't come when command `./blueoil.sh train ...`
```
# supported task types are 'classification', 'object_detection' and 'semantic_segmentation'.
task_type: classification

network_name: LmnetV1Quantize

dataset:
  format: Caltech101
  train_path: lmnet/tests/fixtures/datasets/dummy_classification/
  test_path:

trainer:
  batch_size: 10
  epochs: 10
# supported 'learning_rate_setting' is 'tune1', 'tune2', 'tune3', 'fixed'.
#   'tune1' is 2 times decay, learning rate reduce to 1/10 on epoch/2 and epoch-1.
#   'tune2' is 3 times decay, learning rate reduce to 1/10 on epoch/3 and epoch*2/3 and epoch-1.
#   'tune3' is warmup and 3 times decay, warmup learning rate 1/1000 in 1 epoch, then train same as 'tune2'.
#   'fixed' is constant learning rate.
  learning_rate_setting: tune1
  initial_learning_rate: 0.001

network:
  quantize_first_convolution: no

common:
  image_size:
    - 128  # height
    - 128  # width

  # set pretrain model name. currently, this feature is not supported, always ignored.
  pretrain_model: false

  # enable dataset prefetch, set false if weired problem happens
  dataset_prefetch: true

  data_augmentation:
    - FlipLeftRight:
        - probability: 0.5
    - Hue:
        - value: (-10, 10)
```

Please tell me how the error come ?

**Comment by lm-jira** [comment]
@ruimashita 
My config file is like the one below.

```
# supported task types are 'classification', 'object_detection' and 'semantic_segmentation'.
task_type: classification

network_name: LmnetV1Quantize

dataset:
  format: Caltech101
  train_path: /storage/dataset/Caltech101/101_ObjectCategories/
  test_path:

trainer:
  batch_size: 10
  epochs: 10
# supported 'learning_rate_setting' is 'tune1', 'tune2', 'tune3', 'fixed'.
#   'tune1' is 2 times decay, learning rate reduce to 1/10 on epoch/2 and epoch-1.
#   'tune2' is 3 times decay, learning rate reduce to 1/10 on epoch/3 and epoch*2/3 and epoch-1.
#   'tune3' is warmup and 3 times decay, warmup learning rate 1/1000 in 1 epoch, then train same as 'tune2'.
#   'fixed' is constant learning rate.
  learning_rate_setting: tune1
  initial_learning_rate: 0.001

network:
  quantize_first_convolution: yes

common:
  image_size:
    - 128  # height
    - 128  # width

  # set pretrain model name. currently, this feature is not supported, always ignored.
  pretrain_model: false

  # enable dataset prefetch, set false if weired problem happens
  dataset_prefetch: true

  data_augmentation:
    - Blur:
        - value: (0, 1)
    - Brightness:
        - value: (0.75, 1.25)
    - FlipLeftRight:
        - probability: 0.5
```

Now, I keep the augmenter repository separated from my main repository like this.
```
GitHub
|-blueoil  (main repository)
|-review_pr
       |-blueoil (repository for `add-augmentor-support` branch)
```

Then, I just found that the error comes only when I build the docker from `review_pr/blueoil` before training but the error does not come when I build the docker from `blueoil` before training.

**Comment by ruimashita** [comment]
@lm-jira I understand 👍 
Please give me the error detail, Please copy paste it

**Comment by lm-jira** [comment]
@ruimashita  Here you are
```
Traceback (most recent call last):
  File "blueoil/blueoil_train.py", line 51, in <module>
    main()
  File "/usr/local/pyenv/versions/python3.6/lib/python3.6/site-packages/click/core.py", line 722, in __call__
    return self.main(*args, **kwargs)
  File "/usr/local/pyenv/versions/python3.6/lib/python3.6/site-packages/click/core.py", line 697, in main
    rv = self.invoke(ctx)
  File "/usr/local/pyenv/versions/python3.6/lib/python3.6/site-packages/click/core.py", line 895, in invoke
    return ctx.invoke(self.callback, **ctx.params)
  File "/usr/local/pyenv/versions/python3.6/lib/python3.6/site-packages/click/core.py", line 535, in invoke
    return callback(*args, **kwargs)
  File "blueoil/blueoil_train.py", line 47, in main
    run(config_file, experiment_id)
  File "blueoil/blueoil_train.py", line 29, in run
    run_train(network=None, dataset=None, config_file=lmnet_config_file, experiment_id=experiment_id, recreate=False)
  File "/home/blueoil/lmnet/executor/train.py", line 373, in run
    config = config_util.load(config_file)
  File "/home/blueoil/lmnet/lmnet/utils/config.py", line 118, in load
    config = loader(config_file)
  File "/home/blueoil/lmnet/lmnet/utils/config.py", line 126, in _load_py
    config_module = module_loader.load_module(config_file)
  File "/home/blueoil/lmnet/lmnet/utils/module_loader.py", line 38, in load_module
    module = importlib.import_module(module_path)
  File "/usr/local/pyenv/versions/python3.6/lib/python3.6/importlib/__init__.py", line 126, in import_module
    return _bootstrap._gcd_import(name[level:], package, level)
  File "<frozen importlib._bootstrap>", line 994, in _gcd_import
  File "<frozen importlib._bootstrap>", line 971, in _find_and_load
  File "<frozen importlib._bootstrap>", line 955, in _find_and_load_unlocked
  File "<frozen importlib._bootstrap>", line 665, in _load_unlocked
  File "<frozen importlib._bootstrap_external>", line 674, in exec_module
  File "<frozen importlib._bootstrap_external>", line 781, in get_code
  File "<frozen importlib._bootstrap_external>", line 741, in source_to_code
  File "<frozen importlib._bootstrap>", line 219, in _call_with_frames_removed
  File "/home/blueoil/augment_test.py", line 128
    ])
    ^
SyntaxError: invalid syntax
ERROR: Training exited with a non-zero status
```

**Comment by ruimashita** [comment]
@lm-jira thanks.
I already merge latest master into this branch, Does the error stil come?

**Comment by lm-jira** [comment]
@ruimashita The same error still comes. 
Is it because of the new folder I created?

**Comment by ruimashita** [comment]
@lm-jira 
The docker image is wrong.
When I did `docker build --pull`  which is pull latest base docker image an build again, The error didn't come.

**Comment by ruimashita** [comment]
But I'm not sure which part in docker image is wrong...

**Comment by lm-jira** [comment]
@ruimashita You're right. After adding `--pull` option in `docker_build.sh`, the model works fine.
However, when `Crop` augmentation is selected, the error below occurs.

```
Traceback (most recent call last):
  File "blueoil/blueoil_train.py", line 51, in <module>
    main()
  File "/usr/local/pyenv/versions/python3.6/lib/python3.6/site-packages/click/core.py", line 722, in __call__
    return self.main(*args, **kwargs)
  File "/usr/local/pyenv/versions/python3.6/lib/python3.6/site-packages/click/core.py", line 697, in main
    rv = self.invoke(ctx)
  File "/usr/local/pyenv/versions/python3.6/lib/python3.6/site-packages/click/core.py", line 895, in invoke
    return ctx.invoke(self.callback, **ctx.params)
  File "/usr/local/pyenv/versions/python3.6/lib/python3.6/site-packages/click/core.py", line 535, in invoke
    return callback(*args, **kwargs)
  File "blueoil/blueoil_train.py", line 47, in main
    run(config_file, experiment_id)
  File "blueoil/blueoil_train.py", line 29, in run
    run_train(network=None, dataset=None, config_file=lmnet_config_file, experiment_id=experiment_id, recreate=False)
  File "/home/blueoil/lmnet/executor/train.py", line 373, in run
    config = config_util.load(config_file)
  File "/home/blueoil/lmnet/lmnet/utils/config.py", line 118, in load
    config = loader(config_file)
  File "/home/blueoil/lmnet/lmnet/utils/config.py", line 126, in _load_py
    config_module = module_loader.load_module(config_file)
  File "/home/blueoil/lmnet/lmnet/utils/module_loader.py", line 38, in load_module
    module = importlib.import_module(module_path)
  File "/usr/local/pyenv/versions/python3.6/lib/python3.6/importlib/__init__.py", line 126, in import_module
    return _bootstrap._gcd_import(name[level:], package, level)
  File "<frozen importlib._bootstrap>", line 994, in _gcd_import
  File "<frozen importlib._bootstrap>", line 971, in _find_and_load
  File "<frozen importlib._bootstrap>", line 955, in _find_and_load_unlocked
  File "<frozen importlib._bootstrap>", line 665, in _load_unlocked
  File "<frozen importlib._bootstrap_external>", line 678, in exec_module
  File "<frozen importlib._bootstrap>", line 219, in _call_with_frames_removed
  File "/home/blueoil/augment_test2.py", line 129, in <module>
    Crop(size=128, resize='None', ),
  File "/home/blueoil/lmnet/lmnet/data_augmentor.py", line 224, in __init__
    raise Exception("Expected float or int, tuple/list with 2 entries. Got %s." % (type(resize)))
Exception: Expected float or int, tuple/list with 2 entries. Got <class 'str'>.
ERROR: Training exited with a non-zero status
```

**Comment by lm-jira** [comment]
@ruimashita There are also two more error from two augmenters.

For `RandomErasingForDetection` augmenter, there is an error below.
```
  File "/home/blueoil/lmnet/lmnet/data_augmentor.py", line 658, in __call__
    for i in range(boxes.shape[0]):
AttributeError: 'NoneType' object has no attribute 'shape'
ERROR: Training exited with a non-zero status
```

And for `SSDRandomCrop` augmenter, there is an error below.
```
  File "/home/blueoil/lmnet/lmnet/data_augmentor.py", line 779, in __call__
    if boxes.size == 0:
AttributeError: 'NoneType' object has no attribute 'size'
ERROR: Training exited with a non-zero status
```

**Comment by ruimashita** [comment]
@lm-jira 
Thanks. 
`RandomErasingForDetection` and `SSDRandomCrop` is only for object detection task augmentation.
I will output some error message more user friendly.

**Comment by ruimashita** [comment]
@lm-jira 
I implemented checker of augmentation. https://github.com/blue-oil/blueoil/pull/78/commits/e5f29b9c107fb7a641df5f949f54f4f4e314c898
If the augmentation `SSDRandomCrop` is used in classification task, not available task, raise exception "The SSDRandomCrop augmentation can't be used in IMAGE.CLASSIFICATION task" in train command.
But it is little big changes, to avoid bug I want move it to as new PR.

**Comment by ruimashita** [comment]
@nlpng Review please.

**Comment by nlpng** [comment]
@ruimashita When I run the ./blueoil.sh init, I couldn't get the augmentation option.
Maybe I didn't catch up with the previous conversation, is this the correct behavior?
```
#### Generate config ####
? your model name ():  model
? choose task type  classification
? choose network  LmnetV1Quantize
? choose dataset format  Caltech101
? training dataset path:  /storage/dataset/
? set validataion dataset? (if answer no, the dataset will be separated for training and validation by 9:1 ratio.)  no
? batch size (integer):  10
? image size (integer x integer):  128x128
? how many epochs do you run training (integer):  100
? initial learning rate:  0.001
? choose learning rate setting(tune1 / tune2 / tune3 / fixed):  fixed
? apply quantization at the first layer?  yes

A new configuration file generated: model.yml
  - Your next step is training.
  - You can customize some miscellaneous settings according to the comment.
```

**Comment by ruimashita** [comment]
@nlpng 
probably need to build docker.

**Comment by ruimashita** [comment]
@lm-jira @nlpng When does the review end ?

**Comment by lm-jira** [comment]
@ruimashita 
For me, this PR is okay now.
The error from `Crop` augmentor is gone.

**Comment by ruimashita** [comment]
@lm-jira 
Thanks 😄 Please `approve` this PR. No one approve this PR, I can't merge this PR.
https://help.github.com/articles/approving-a-pull-request-with-required-reviews/

**Pull request merged**